src/_pytest/python.py

```diff
@@ -953,7 +953,7 @@ def _idval_from_function(self, val: object, argname: str, idx: int) -> str | Non
         if self.idfn is None:
             return None
         try:
-            id = self.idfn(val)
+            id = self.idfn(val, argname, idx)
```

Comment:
This is a breaking change,the old signature needs to be supported as well 

I recommend using a callable protocol as starting points